src/mappings/nabla.ts

```diff
@@ -34,35 +42,77 @@ function decodeEvent(
     }
 }
 
+async function isSwapPoolEvent(ctx: EventHandlerContext) {
+    try {
+        const contract = new swapPoolAbi.Contract(ctx, ctx.event.args.contract)
+        await contract.router()
+        await contract.backstop()
+        await contract.accumulatedSlippage()
+        await contract.poolCap()
+        return true
+    } catch {
+        return false
+    }
+}
+
+async function isRouterEvent(ctx: EventHandlerContext) {
+    try {
+        const contract = new routerAbi.Contract(ctx, ctx.event.args.contract)
+        await contract.poolByAsset(new Uint8Array(32))
+        return true
+    } catch {
+        return false
+    }
+}
+
+async function isBackstopPoolEvent(ctx: EventHandlerContext) {
+    try {
+        const contract = new backstopPoolAbi.Contract(
+            ctx,
+            ctx.event.args.contract
+        )
+        await contract.getBackedPool(0n)
+        return true
+    } catch {
+        return false
+    }
+}
+
+async function verifyEvent(veryfier: Function, ctx: EventHandlerContext) {
+    return veryfier(ctx)
+}
+
 // Iterates over all decoders and returns the first successfully decoded event
-function getEventAndEventType(ctx: EventHandlerContext): {
+async function getEventAndEventType(ctx: EventHandlerContext): Promise<{
     event: Event | null
     eventType: EventType | null
-} {
-    const decoders = [erc20Abi, backstopPoolAbi, swapPoolAbi, routerAbi]
+}> {
+    const veryfiers = [isBackstopPoolEvent, isRouterEvent, isSwapPoolEvent]
+    const decoders = [backstopPoolAbi, routerAbi, swapPoolAbi]
     const eventTypes = [
-        null,
         EventType.BackstopPoolEvent,
-        EventType.SwapPoolEvent,
         EventType.RouterEvent,
+        EventType.SwapPoolEvent,
     ]
 
-    for (let i = 0; i < decoders.length; i++) {
-        const { result: event, error: err } = decodeEvent(
-            ctx.event.args.data,
-            decoders[i]
-        )
-        if (!err) {
-            const eventType = eventTypes[i]
-            return { event, eventType }
+    for (let i = 0; i < veryfiers.length; i++) {
+        if (await verifyEvent(veryfiers[i], ctx)) {
```

Comment:
Nit: We can simplify this logic now. The implementations of the `verifiers` that are called as part of the `verifyEvent` function do decode the contracts. Thus, the call to `decodeEvent()` in ll. 100-102 is actually redundant _if_ we pass the decoded event out of the verifiers. 

We could change the return type of the verifier functions to the event instead of a boolean. And in l. 99 we then check if the returned `event !== undefined`. However, changing it this way might not necessarily make the code easier to read/understand.

Maybe it's enough if we just add a comment here. WDYT @adelarja?
```suggestion
    // Iterate over all verifiers and try to decode the event to the given type
    for (let i = 0; i < verifiers.length; i++) {
        if (await verifyEvent(verifiers[i], ctx)) {
```

Reply:
> Nit: We can simplify this logic now. The implementations of the `verifiers` that are called as part of the `verifyEvent` function do decode the contracts. Thus, the call to `decodeEvent()` in ll. 100-102 is actually redundant _if_ we pass the decoded event out of the verifiers.
> 
> We could change the return type of the verifier functions to the event instead of a boolean. And in l. 99 we then check if the returned `event !== undefined`. However, changing it this way might not necessarily make the code easier to read/understand.
> 
> Maybe it's enough if we just add a comment here. WDYT @adelarja?

@ebma thanks for the comment.

I totally agree with you, we don't need to call decodeEvent once we know the event type.

While I was improving the logic, something came to my mind:

If in the future the contracts are upgraded, and for example, new events are emitted, or the function names change, or whatever. In that case we would have two choices:
1. Track only events emitted by the upgraded smart contracts, updating the ABIs.
2. Track both, the events emitted by the upgraded contracts and by the old smart contracts (using both the old and the updated ABIs).

It is not urgent, but it might be good to keep it in mind for the future.

Reply:
That's a good point but I think practically speaking we will not face this issue in the future. The reason being that smart contracts are not upgradable, at least not without putting in some extra work, which AFAIK Nabla is not planning to add. So once Nabla launches on Pendulum, the smart contracts will probably (or ideally) stay like that forever and we don't have to worry about ABIs changing over time.